### PR TITLE
bkh - add functionality to delete images

### LINKED ISSUE #8: Delete Image
#### As a user I want to delete an image.

__Story__:

Our users are awesome at adding images to our web service. Some of our users
add dozens in a single session. Occasionally, however, some users add the wrong
image, and would like a way to remove those images. Please provide our users
with the ability to delete images. Note that at this point we trust our users
not to abuse this feature and delete more images than they should.

__Acceptance criteria__:
- [x] The user is prompted for confirmation prior to deletion.
- [x] Deleted images are removed from the homepage.

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by BinKuangHsiehApp**
Closes #8

### CODE REVIEW

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -53,6 +54,14 @@ def test_should_get_index__with_images_last_in_first_out
     end
   end
 
+  def test_should_delete_image
+    new_image = create_image
+    create_image
```
Comment (D-D-M) [inline test/controllers/images_controller_test.rb]: This one was a little bit tough for me to understand the first time I read it. If I understand the purpose of this correctly, perhaps it could be a little bit clearer like this:
`2.times do { create_image }`
followed by
`delete image_path(Image.first)`

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -22,6 +22,12 @@ def show
     @current_image = Image.find(params[:id])
   end
 
+  def destroy
+    Image.find(params[:id]).destroy
```
Comment (D-D-M) [inline app/controllers/images_controller.rb]: I'll cook up a scenario/riddle for you here:

User opens "show" page for an image in a new tab. User goes back to the old tab and opens the same image in an even newer tab. Now they've got 2 tabs open with the same image. User clicks delete in the first tab. What happens if they try to delete the image in the second tab? You should try this in your local environment to see what happens :)

What should happen is that `Image.find` will fail to find anything, and it'll throw some kind of `RecordNotFound` exception. This is pretty easy to fix though!

Option 1:
You can `rescue` the exception in the controller action, set a flash message that says "Unable to delete image", and redirect to images_path.

Option 2:
You can do `image = Image.find_by(id: params[:id])`. If `image` is nil, you can set the flash message to "Unable to delete image", and redirect to the images_path.

Both options are fine. The only difference is that `find` will throw an exception if it fails, whereas `find_by` will return `nil` if it fails.

You'd probably also want to add a test for this edge case as well.

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (D-D-M) [inline test/controllers/images_controller_test.rb]: And actually, now that I've thought about it some more, I'm not sure that you even need to create a second image – all you want to know is that your `delete` call took the count from 1 to 0, so you can probably get rid of line 59 entirely. What do you think?

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -19,7 +19,24 @@ def index
   end
 
   def show
-    @current_image = Image.find(params[:id])
+    image_or_nil = Image.find_by(id: params[:id])
```
Comment (D-D-M) [inline app/controllers/images_controller.rb]: I like this overall approach, but I don't think it's necessary to name the variable by every value it might be. With this name, the `if` statement below looks like it's asking if the current variable is "image or nil", and the answer to that is always going to be yes 😂  So just `image = Image.find_by` is probably fine.

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -19,7 +19,24 @@ def index
...
+    if image_or_nil
+      @current_image = image_or_nil
+    else
+      flash[:danger] = 'Unable to access image.'
+      redirect_to images_path
+    end
+  end
+
+  def destroy
+    image_or_nil = Image.find_by(id: params[:id])
```
Comment (D-D-M) [inline app/controllers/images_controller.rb]: Same as above.

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (BinKuangHsiehApp) [inline test/controllers/images_controller_test.rb]: I wanted to check if the tag count was also reduced. but I think you may be right, the second one probably isn't necessary

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -53,6 +54,34 @@ def test_should_get_index__with_images_last_in_first_out
...
 
+  def test_should_delete_image
+    new_image = create_image(tag_list: 'tag2, tag_delete')
+    create_image
+    assert_difference 'Image.count', -1 do
+      assert_difference 'ActsAsTaggableOn::Tag.count', -1 do
+        delete image_path(new_image)
+      end
+    end
+    get images_path
```
Comment (D-D-M) [inline test/controllers/images_controller_test.rb]: It's not common for a functional test like this to make multiple requests (multiple requests are more in the domain of integration tests). The idea is that you want to test one request, and make sure that that request does everything you expect it to do. The more extraneous requests you involve in a test (`get images_path`), the more difficult it will be to fix the tests when you change the controller action that the path aims at (i.e. `ImagesController#index`).

So instead, you can make your `delete` request, and then use [assert_redirected_to](https://apidock.com/rails/ActionDispatch/Assertions/ResponseAssertions/assert_redirected_to), to make sure that you're being redirected to the index page after you've deleted an image.

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -53,6 +54,34 @@ def test_should_get_index__with_images_last_in_first_out
...
+    assert_select '.alert-success', text: 'Image deleted'
```
Comment (D-D-M) [inline test/controllers/images_controller_test.rb]: The above comment makes this tough – if you're not going back to the index page, how are you supposed to see that the flash message was successful? Luckily, you can access the `flash` hash from inside these controller tests.

Something like `assert_equal 'Image deleted', flash[:success]' should work without having to make a GET request to images_path :+1:

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -53,6 +54,34 @@ def test_should_get_index__with_images_last_in_first_out
...
+  end
+
+  def test_should_delete_image__cannot_access_from_index_again
+    new_image = create_image(tag_list: 'tag2, tag_delete')
+    delete image_path(new_image)
+    get image_path(new_image)
+    get images_path
```
Comment (D-D-M) [inline test/controllers/images_controller_test.rb]: Again, you don't need these extra requests – just make an assertion on the `flash` hash being equal to whatever you want.

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -53,6 +54,34 @@ def test_should_get_index__with_images_last_in_first_out
...
+    assert_select '.alert-danger', text: 'Unable to access image.'
+  end
+
+  def test_should_delete_image__should_not_delete_again
+    new_image = create_image
+    delete image_path(new_image)
+    delete image_path(new_image)
```
Comment (D-D-M) [inline test/controllers/images_controller_test.rb]: This is one way to delete an image that you're 100% sure won't exist, but to get away from the need for multiple DELETE requests, a common pattern is to do something like this instead:

(You can be certain that there will never be an image in the database with a negative ID. So the test becomes a lot simpler!)

```
assert_difference 'Image.count', 0 do
  delete image_path(id: -1)
end
assert_equal 'Unable to delete image.', flash[:whatever]
```

File: test/controllers/images_controller_test.rb (hunk 7 above)
Comment (D-D-M) [inline test/controllers/images_controller_test.rb]: Actually, this test isn't really about deletion – it seems like it should be written as a test of the `ImagesController#show` action. You really just want to make sure that `get image_path(id: -1)` will do the correct redirect, and display the correct flash message.

File: test/controllers/images_controller_test.rb [hunk 9]
```diff
@@ -53,6 +60,22 @@ def test_should_get_index__with_images_last_in_first_out
...
+    create_image
+    assert_difference 'Image.count', -1 do
+      delete image_path(new_image)
+    end
+    assert_redirected_to images_path
+    assert_equal 'Image deleted', flash[:success]
+  end
+
+  def test_should_delete_image__should_not_delete_again
+    delete image_path(id: -1)
```
Comment (D-D-M) [inline test/controllers/images_controller_test.rb]: You could wrap this in a `assert_difference 'Image.count', 0 do` to make the test more robust, but other than that, this is great 👍
